indexing-service/src/main/java/org/apache/druid/indexing/firehose/IngestSegmentFirehoseFactory.java

```diff
@@ -251,6 +285,179 @@ private long jitter(long input)
+    // same interval).
+    Interval lastInterval = null;
+    for (Interval interval : timeline.keySet()) {
+      if (lastInterval != null) {
+        if (interval.overlaps(lastInterval)) {
+          throw new IAE(
+              "Distinct intervals in input segments may not overlap: [%s] vs [%s]",
+              lastInterval,
+              interval
+          );
+        }
+      }
+      lastInterval = interval;
+    }
+
+    return new ArrayList<>(timeline.values());
+  }
+
+  private void initializeSplitsIfNeeded()
+  {
+    if (splits != null) {
+      return;
+    }
+
+    // isSplittable() ensures this is only called when we have an interval.
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = getTimelineForInterval();
+
+    // We do the simplest possible greedy algorithm here instead of anything cleverer. The general bin packing
+    // problem is NP-hard, and we'd like to get segments from the same interval into the same split so that their
+    // data can combine with each other anyway.
+
+    List<InputSplit<List<WindowedSegmentId>>> newSplits = new ArrayList<>();
+    List<WindowedSegmentId> currentSplit = new ArrayList<>();
+    Map<DataSegment, WindowedSegmentId> windowedSegmentIds = new HashMap<>();
+    long bytesInCurrentSplit = 0;
+    for (TimelineObjectHolder<String, DataSegment> timelineHolder : timelineSegments) {
+      for (PartitionChunk<DataSegment> chunk : timelineHolder.getObject()) {
+        final DataSegment segment = chunk.getObject();
+        final WindowedSegmentId existingWindowedSegmentId = windowedSegmentIds.get(segment);
+        if (existingWindowedSegmentId != null) {
+          // We've already seen this segment in the timeline, so just add this interval to it. It has already
+          // been placed into a split.
+          existingWindowedSegmentId.getIntervals().add(timelineHolder.getInterval());
+        } else {
+          // It's the first time we've seen this segment, so create a new WindowedSegmentId.
+          List<Interval> intervals = new ArrayList<>();
+          // Use the interval that contributes to the timeline, not the entire segment's true interval.
+          intervals.add(timelineHolder.getInterval());
+          final WindowedSegmentId newWindowedSegmentId = new WindowedSegmentId(segment.getId().toString(), intervals);
+          windowedSegmentIds.put(segment, newWindowedSegmentId);
+
+          // Now figure out if it goes in the current split or not.
+          final long segmentBytes = segment.getSize();
+          if (bytesInCurrentSplit + segmentBytes > maxInputSegmentBytesPerTask && !currentSplit.isEmpty()) {
+            // This segment won't fit in the current non-empty split, so this split is done.
+            newSplits.add(new InputSplit<>(currentSplit));
+            currentSplit = new ArrayList<>();
+            bytesInCurrentSplit = 0;
+          }
+          if (segmentBytes > maxInputSegmentBytesPerTask) {
```

Comment:
Hmm, this might not be what we want. The idea of size-based split is to process roughly the same amount of data in each subTask. How ingestSegmentFirehose actually works is, 1) per segment, it reads the timestamp column and 2) reads other columns if the timestamp falls in the given interval (`QueryableIndexStorageAdapter.CursorSequenceBuilder`). 

But, this `if` clause compares the entire segment size with `maxInputSegmentBytesPerTask` even if the interval covers only a part of a segment. I guess we may need to estimate the segment size to be read roughly by computing based on the ratio of the window interval to the whole interval of a segment?

Reply:
I'm happy to implement this today, but as a couple counterpoints before I do so:

- Size-based split does help keep subtasks equal, but it also has the purpose of bounding the amount of disk space needed for each task (because the current implementation downloads all segments at once at the beginning of the task), at least as long as all individual segments fall below the boundary.  It would be a shame if a single task ended up requiring an enormous amount of disk space (and download time) if it ended up being assigned a large number of segments which only contribute a small fraction of their data to the output segment.

- In practice I expect that most people's segments have uniform granularity (or at least are non-overlapping, eg older data having uniform day granularity and newer data having uniform hour granularity) and so the only time that windows will not be the full segment size is at the beginning and end of the total interval.  And even having misalignment there will be a mistake if you're using this to re-ingest data into itself with `appendToSegments: false` since it will drop data in the start and end segments that are outside of the interval.  So I think windowing is not going to be a super common use case in practice, and so doing a performance optimization that only matters for the windowing use case might not be worth the extra code complexity and more complicated documentation needed for `maxInputSegmentBytesPerTask`.  (Of course the overlapping segment use case needs to be supported for *correctness* because that's how the Druid data model works. But that doesn't mean it needs performance optimizations?)

Reply:
> In practice I expect that most people's segments have uniform granularity (or at least are non-overlapping, eg older data having uniform day granularity and newer data having uniform hour granularity) and so the only time that windows will not be the full segment size is at the beginning and end of the total interval. And even having misalignment there will be a mistake if you're using this to re-ingest data into itself with appendToSegments: false since it will drop data in the start and end segments that are outside of the interval. So I think windowing is not going to be a super common use case in practice, and so doing a performance optimization that only matters for the windowing use case might not be worth the extra code complexity and more complicated documentation needed for maxInputSegmentBytesPerTask. (Of course the overlapping segment use case needs to be supported for correctness because that's how the Druid data model works. But that doesn't mean it needs performance optimizations?)

If I understand correctly, windowing would happen depending on the value of `maxInputSegmentBytesPerTask` rather than segment granularity because split is created whenever `bytesInCurrentSplit + segmentBytes` is larger than `maxInputSegmentBytesPerTask`. Am I missing something?

> Size-based split does help keep subtasks equal, but it also has the purpose of bounding the amount of disk space needed for each task (because the current implementation downloads all segments at once at the beginning of the task), at least as long as all individual segments fall below the boundary. It would be a shame if a single task ended up requiring an enormous amount of disk space (and download time) if it ended up being assigned a large number of segments which only contribute a small fraction of their data to the output segment.

Good point. I think it would be an issue, but maybe it would be better to handle it in `IngestSegmentFirehoseFactory` like lazily downloading segments whenever needed (probably worth to combine with `PrefetchableTextFilesFirehoseFactory`?). But, this should be done in a separate PR and I'm fine with going as it is. Would you please add a comment about this concern?

Reply:
> If I understand correctly, windowing would happen depending on the value of maxInputSegmentBytesPerTask rather than segment granularity because split is created whenever bytesInCurrentSplit + segmentBytes is larger than maxInputSegmentBytesPerTask. Am I missing something?

Individual input segments are never split up over multiple tasks. The only times that the window on a segment isn't that segment's full interval are when the segment goes outside of the overall task's boundaries or if part of the segment is overshadowed.  So my point in that paragraph was that in practice, the ratio you're describing will generally be 100%, and I'm not sure the added complexity (including making the meaning of `maxInputSegmentBytesPerTask` more complex) are worth it.

I will add a comment about lazily downloading segments. I suppose there's something positive about failing fast if segments can't be found, but that's probably not that important.

Reply:
Ok, it makes sense.